### DISCUSSION TITLE
Segmentation fault using Linux Docker Container: Windows host

### DISCUSSION THREAD

**Discussion by jmythms** [Kitware/trame#532] [Trame/ParaView] [answered]
I am running into a segmentation fault while running the trame-tutorial/05_paraview/SimpleCone.py within a Linux container, on a Windows host machine. 

![image](https://github.com/Kitware/trame/assets/45446967/8bfd6f84-e564-4f19-a305-31537c9ad8ca)

Steps I followed to error: 
0. Built Container and accessed container bash. 
1. Installed Paraview: downloaded and unzipped into /Paraview.
2. Followed the virtual environment creation in https://kitware.github.io/trame/guide/tutorial/paraview.html
3. used pvpython to run the example within the container

I was able to run the example fine on the Windows host. It also works ok on a Linux host. 

Any suggestions on how/where to debug? Is there another preferred way to run the program within docker? 

Ubuntu version in container:  --platform=linux/x86_64 ubuntu:22.04
Paraview version: ParaView-5.12.0-MPI-Linux-Python3.10-x86_64.tar.gz
Trame version 3.6.2.

**Comment by jmythms** [comment]
Also:
On my Windows host, it seems that the graphics card is being used, and I don't see that message in my Docker container. Wondering if this is contributing to the segfault somehow. 

![image](https://github.com/Kitware/trame/assets/45446967/8c6975b1-da90-4d9c-adb0-a1e37d60ab13)

**Comment by jourdain** [comment]
1. you need an offscreen version of ParaView to work inside docker. (EGL if you want to use your GPU)
2. you need a docker image with nvidia drivers
3. you need to run your docker with something like `--gpus all` 

Then with trame, you can use `kitware/trame:py3.10-glvnd` image, and configure your app to run using pvpython like we do [here](https://github.com/Kitware/paraview-visualizer/blob/master/docker/setup/apps.yml#L4-L18).

**Reply by jmythms** [reply]
Thank you. If without using GPUs, use the osmesa version and skip steps 2 and 3?

**Reply by jourdain** [reply]
yes and you can use any trame base image

**Reply by jmythms** [reply]
Still running into the same error :(

I also tried to run https://github.com/Kitware/trame/tree/master/examples/07_paraview/Docker. I updated the Paraview link to the latest version. Even having trouble with that example.  

![image](https://github.com/Kitware/trame/assets/45446967/6c91a9df-4017-4fd1-9ebc-88852d9d1ff8)

![image](https://github.com/Kitware/trame/assets/45446967/9c2aa1fe-b5cd-4377-bb1b-4169604ed4ec)

**Reply by jourdain** [reply, marked as answer]
Actually 5.12 has an issue with osmesa (totally forgot about it).

Try grabbing the egl one. Even if you don't have a gpu, it should still work by using the mesa implementation of egl.

**Reply by jmythms** [reply]
I think I finally have it working. Thanks!

**Reply by jmythms** [reply]
One more (maybe dumb) question, is it possible to use GPUs other than nvidia? As in can I set Docker up so that Paraview will use any graphics card that is available?

**Reply by jourdain** [reply]
FYI, 5.12 osmesa still link against libX11. So the system needs that lib or it will crash.

**Comment by jmythms** [comment]
Sorry to keep asking, but can I implement the version of kitware/trame:py3.10-glvnd in a plain ubuntu base image? I can't see the layers in that image, and wanted to keep things open if possible. Or, to put it differently, is there another way to get the dependencies needed for GPU other than using the kitware/trame:py3.10-glvnd base image?

**Reply by jourdain** [reply]
Yes you can use the same base image as we do [here](https://github.com/Kitware/trame/blob/master/.github/workflows/docker.yml#L73).
